File: src/bmsx/machine/devices/input/sample_latch.ts
import { Input } from '../../../input/manager';
import { InputControllerActionTable } from './action_table';
import { InputControllerEventFifo } from './event_fifo';

export type InputControllerSampleLatchState = {
	sampleArmed: boolean;
	sampleSequence: number;
	lastSampleCycle: number;
};

export class InputControllerSampleLatch {
	private sampleArmed = false;
	private sampleSequence = 0;
	private lastSampleCycle = 0;

	public constructor(
		private readonly input: Input,
		private readonly actionTable: InputControllerActionTable,
		private readonly eventFifo: InputControllerEventFifo,
	) {}

	public reset(): void {
		this.sampleArmed = false;
		this.sampleSequence = 0;
		this.lastSampleCycle = 0;
	}

	public arm(): void {
		this.sampleArmed = true;
	}

	public cancel(): void {
		this.sampleArmed = false;
	}

	public onVblankEdge(currentTimeMs: number, nowCycles: number): void {
		if (!this.sampleArmed) {
			return;
		}
		this.sampleSequence = (this.sampleSequence + 1) >>> 0;
		this.lastSampleCycle = nowCycles >>> 0;
		this.input.samplePlayers(currentTimeMs);
		this.actionTable.sampleCommittedActions(this.eventFifo);
		this.sampleArmed = false;
	}

	public captureState(): InputControllerSampleLatchState {
		return {
			sampleArmed: this.sampleArmed,
			sampleSequence: this.sampleSequence,
			lastSampleCycle: this.lastSampleCycle,
		};
	}

	public restoreState(state: InputControllerSampleLatchState): void {
		this.sampleArmed = state.sampleArmed;
		this.sampleSequence = state.sampleSequence;
		this.lastSampleCycle = state.lastSampleCycle;
	}
}

File: src/bmsx_cpp/machine/devices/input/sample_latch.cpp
#include "machine/devices/input/sample_latch.h"

#include "input/manager.h"
#include "machine/devices/input/action_table.h"
#include "machine/devices/input/event_fifo.h"

namespace bmsx {

InputControllerSampleLatch::InputControllerSampleLatch(Input& input, InputControllerActionTable& actionTable, InputControllerEventFifo& eventFifo)
	: m_input(input)
	, m_actionTable(actionTable)
	, m_eventFifo(eventFifo) {
}

void InputControllerSampleLatch::reset() {
	m_sampleArmed = false;
	m_sampleSequence = 0u;
	m_lastSampleCycle = 0u;
}

void InputControllerSampleLatch::arm() {
	m_sampleArmed = true;
}

void InputControllerSampleLatch::cancel() {
	m_sampleArmed = false;
}

void InputControllerSampleLatch::onVblankEdge(f64 currentTimeMs, u32 nowCycles) {
	if (!m_sampleArmed) {
		return;
	}
	m_sampleSequence += 1u;
	m_lastSampleCycle = nowCycles;
	m_input.samplePlayers(currentTimeMs);
	m_actionTable.sampleCommittedActions(m_eventFifo);
	m_sampleArmed = false;
}

InputControllerSampleLatchState InputControllerSampleLatch::captureState() const {
	InputControllerSampleLatchState state;
	state.sampleArmed = m_sampleArmed;
	state.sampleSequence = m_sampleSequence;
	state.lastSampleCycle = m_lastSampleCycle;
	return state;
}

void InputControllerSampleLatch::restoreState(const InputControllerSampleLatchState& state) {
	m_sampleArmed = state.sampleArmed;
	m_sampleSequence = state.sampleSequence;
	m_lastSampleCycle = state.lastSampleCycle;
}

} // namespace bmsx
